Checksum verification: add content

3mkusiak opened the pull request:
Add content for checksum verification.

dancesWithMachines committed:
Checksum verification: add content

Add content for checksum verification.

Signed-off-by: Mateusz Kusiak <mateusz.kusiak@3mdeb.com>

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
 
+Checksum as a concept, is a redundancy check for detecting variations in data.
+Checksums are created as result of calculating binary values of a block or
+packet of data using sophisticated algorithms. Comparing checksums allows for
```
> sophisticated algorithms

parity checksum == xor all bits
It can use sophisticated algorithms but there are also very simple ones.

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+Checksum as a concept, is a redundancy check for detecting variations in data.
+Checksums are created as result of calculating binary values of a block or
+packet of data using sophisticated algorithms. Comparing checksums allows for
+verifying data integrity, yet is truthfulness is dependent of algorithm
```
```suggestion
verifying data integrity, yet its truthfulness is dependent of algorithm
```

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+verifying data integrity, yet is truthfulness is dependent of algorithm
+used[^1].
+
+Checksums values come in form of unique, fixed string of characters referred to
+as hash. Exact form of hash is dependent on algorithm used. Checksum algorithms
+ensure even tiniest changes made to a block of data, would result in checksum
+value changing completely. This property makes them easy to compare without
+need for any additional components. Common algorithms include: sha-256, md5 or
+sha-1[^2].
```
> even tiniest changes made to a block of data, would result in checksum
value changing completely. This property

I think that's property of good cryptographic hash function (I think my professor called it mixing property i.e. small input change results in big output change).

> Checksums values come in form of unique

I'm not sure what you mean here. It's impossible for checksums (or hashes) to be unique unless space of input values is <= space of output values.

> fixed string of characters
Checksum can be integer value e.g. IPv4 header checksum which is sum of all 2-byte values in header (though I guess you could view them as string characters (unless they are unprintable))

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+
+#### Checksum verification in chain of trust context
+
+Checksum verification can be used as a part of chain of trust mechanism.
```
I think any chain of trust would require use of cryptographic hash functions (which have more stringent properties) instead of any checksum.

m-iwanicki reviewed (changes requested):
This whole section mixes checksums, hash functions and cryptographic hash functions.
Each concept is different (more or less even if all of them are related)

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+#### debsums
+
+A supplementary integrity checks can also be performed on OS level.
+`debsums` is an utility for verifying debian package files against their
```
```suggestion
`debsums` is a utility for verifying Debian package files against their
```

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+Checksum verification can be used as a part of chain of trust mechanism.
+Particular example is Google's Verified Boot found on Android based devices.
+Verified boot cryptographically verifies executable code and data as the device
+boots. It does so for both files and partitions using multiple mechanisms, one
```
```suggestion
boots. It does so for both files and partitions using multiple mechanisms,
```

m-iwanicki commented on verified_boot_main.md:
```diff
@@ -70,6 +70,42 @@
+
+#### References
+
+[^1] [Checksum Definition](https://www.linfo.org/checksum.html)
+[^2] [Exploring Checksums: Types, Uses, And Verification Methods](https://linuxsecurity.com/features/what-are-checksums-why-should-you-be-using-them)
+[^3] [Verify Boot](https://source.android.com/docs/security/features/verifiedboot/verified-boot)
+[^4] [debsums](https://manpages.ubuntu.com/manpages/trusty/man1/debsums.1.html)
```
```suggestion
[^1]: [Checksum Definition](https://www.linfo.org/checksum.html)
[^2]: [Exploring Checksums: Types, Uses, And Verification Methods](https://linuxsecurity.com/features/what-are-checksums-why-should-you-be-using-them)
[^3]: [Verify Boot](https://source.android.com/docs/security/features/verifiedboot/verified-boot)
[^4]: [debsums](https://manpages.ubuntu.com/manpages/trusty/man1/debsums.1.html)
```

3mkusiak commented:
I agree I mixed cryptographic hash functions, I'll fix it.

dancesWithMachines committed:
Checksum verification: separate definitions

Hash, checksum and hash algorithms were mixed up and inconsistent.
Separate and clear up definitions and connections between them.

Signed-off-by: Mateusz Kusiak <mateusz.kusiak@3mdeb.com>

3mkusiak commented:
@m-iwanicki I worked on clearing up definitions and reordering concept to hopefully clear out how they are related. I did not dig into checksums (the ip example) as I don't think that's the scope here. Let me know whatcha think.

macpijan commented:
@m-iwanicki @3mkusiak I see some back and forth on definitions here. I suggest that both of you go through the https://github.com/3mdeb/verified-boot/pull/7 at least briefly.

If there are some missing definitions we need, please suggest them to @philipanda

ArrayBolt3 commented on verified_boot_main.md:
```diff
@@ -70,6 +70,61 @@
+#### Common hashing algorithms
+
+Common hashing algorithms include[^3]:
+* MD-5 - created in 1991, but now considered compromised as it was discovered
+how to decode it.
```
I don't think "decode" is the right term here, since MD5 isn't a reversible encryption algorithm. What has been discovered are successful collision attacks against MD5, so maybe this would be more accurate:

```suggestion
* MD-5 - created in 1991, but now considered compromised as successful collision attacks
against it have been demonstrated.
```

https://web.archive.org/web/20170517115509/http://www.win.tue.nl/hashclash/On%20Collisions%20for%20MD5%20-%20M.M.J.%20Stevens.pdf and https://web.archive.org/web/20170325033522/http://www.win.tue.nl/hashclash/rogue-ca/ can be used as references for this.

3mkusiak replied:
Fixed.

ArrayBolt3 commented on verified_boot_main.md:
```diff
@@ -70,6 +70,61 @@
+Common hashing algorithms include[^3]:
+* MD-5 - created in 1991, but now considered compromised as it was discovered
+how to decode it.
+* SHA family - developed by US government. A recent SHA-3 version uses
+randomness in the code making it more difficult to decode. Currently an
+industry standard.
```
What does "randomness in the code" mean? This would probably benefit from a reference. Similarly, "decode" may need replaced with "find collisions for".

3mkusiak replied:
Fixed and added reference.

ArrayBolt3 reviewed (commented):
Left a couple notes in case they're helpful.

adrelanos commented on verified_boot_main.md:
```diff
@@ -70,6 +70,61 @@
+
+#### Extending chain of trust to OS
+
+A supplementary integrity checks can also be performed on OS level.
```
Good to insert newline here?

adrelanos commented on verified_boot_main.md:
```diff
@@ -70,6 +70,61 @@
+checksums. The MD-5 algorithm is used in the process. `debsums` is primarily
+intended for determining if installed files were modified or damaged. It also
+serves limited usage as a security tool[^5].
+
```
debsums isn't great of this yet indeed. But that should not be understood as discouraging.

Worth mentioning other tools? Worth mentioning that this is not a inherent issue and fixable in theory?

Mention debcheckroot does exist and that other similar tools might exist?

3mkusiak replied:
>Worth mentioning other tools? 

Sure.

>Worth mentioning that this is not a inherent issue and fixable in theory?

IMO, we should focus on the state it is in currently. My take is that wiki should not "wonder about the future".

> Mention debcheckroot does exist and that other similar tools might exist

Ack.

3mkusiak replied:
Added `debcheckroot` and `AIDE`.

dancesWithMachines committed:
Checksum verification: add more OS level tools

Add few more system level tools as examples on extending chain of trust
to OS level. Fix issues regarding hash algorithms.

Signed-off-by: Mateusz Kusiak <mateusz.kusiak@3mdeb.com>

macpijan reviewed (approved):
I can see @3mkusiak has fixed the remaining threads.
I will keep them open just in case original authors do not agree and need further clarification.

We can always fix that in a followup if needed.
